Optimize scatter/gather kernel for ARM.

maajidkhann opened the pull request:
**Optimize scatter/gather kernel for ARM.**

*Retain FBGEMM kernel for x86.
*Unlocking expanded-index fast path and major speedup on ARM CPU's.

**Performance Evaluation:**
**Hardware used:** Graviton 3 - 32 core
 **Package versions:**
  torch              v2.7.0
  torch_geometric    v2.4.0

**Testing:**
Benchmarked the optimized scatter/gather kernel implementation on an ARM-based 
Graviton3 server using a graph neural network inference workload with PyTorch Geometric. 
Loaded 100 real-world graph samples and performed multiple warmup and timed inference 
passes using a pre-trained model. Profiling was enabled via torch.profiler (CPU-only) 
to capture per-op performance metrics. The profiler traces and MSE accuracy 
confirmed both functional correctness and significant speedup for 
scatter/gather-heavy workloads on ARM.




**Total Evaluation Time:**

| **OMP\_NUM\_THREADS / Cores** | **Original (v2.7.0) Time** | **fbgemm-patched (v2.7.0 + patch) Time** |
| ----------------------------- | -------------------------- | ---------------------------------------- |
| 16 / 32                       | 169.58s                    | 144.23s                                  |
| 32 / 32                       | 165.42s                    | 138.70s                                |

**Summary:**
With the fbgemm patch, total evaluation time decreased from 169.58s to 144.23s (a 15% reduction) with 16 threads, and from 165.42s to 138.70s (a 16% reduction) with 32 threads.
This highlights a consistent end-to-end inference speedup across thread counts on ARM.
____________________________________________________________________________________________________________________________

**PyTorch profiler Data:**
| Version                   | OMP\_NUM\_THREADS | Self CPU % | Self CPU Time (ms) |
| ------------------------- | ----------------- | ---------- | ------------------ |
| **Official v2.7.0**       | 16                | 21.15%     | 365.06             |
| **Official v2.7.0**       | 32                | 21.03%     | 353.87             |
| **v2.7.0 + fbgemm patch** | 16                | 10.32%     | 153.30             |
| **v2.7.0 + fbgemm patch** | 32                | 5.70%      | 79.31              |


__________________________________________________________________________________________________________________________________

**Total time taken by the kernel (In ms):**
| OMP\_NUM\_THREADS | Official Time (ms) | Patched Time (ms) | Speedup (×) | Time Reduction (%) |
| ----------------- | ------------------ | ----------------- | ----------- | ------------------ |
| 16                | 365.06             | 153.30            | **2.38×**   | **57.99%**         |
| 32                | 353.87             | 79.31             | **4.46×**   | **77.59%**         |

**Summary:**
With OMP_NUM_THREADS=16, the fbgemm patch reduced scatter_add_ CPU time by ~58% (2.38× faster).

With OMP_NUM_THREADS=32, the patch reduced scatter_add_ CPU time by ~78% (4.46× faster).

The patch also reduced the relative CPU usage of scatter kernels (Self CPU %) substantially.


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @snadampal @milpuz01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia @malfet

maajidkhann committed:
Optimize scatter/gather kernel for ARM.

Retain FBGEMM kernel for x86, unlocking expanded-index
fast path and major speedup on ARM CPU's.

Signed-off-by: maajidkhann <maajidkhan.n@fujitsu.com>

maajidkhann commented:
@pytorchbot label "module: arm"

maajidkhann commented:
@pytorchbot label "ciflow/linux-aarch64"

maajidkhann commented:
@pytorchbot label "topic: performance"

fadara01 commented:
Thank you for your PR!
It would be nice if you could add a benchmarking script that shows the speedups you're reporting (it can just be a single op running on dummy but representative data)

fadara01 reviewed (changes requested):
could you also please elaborate on the test plan for this change?
do we have tests already exercising this path for AArch64?

maajidkhann committed:
Minor Fixes Added

Signed-off-by: maajidkhann <maajidkhan.n@fujitsu.com>

maajidkhann commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -1997,20 +1993,21 @@ Tensor index_fill(
   return self.clone(at::MemoryFormat::Preserve).index_fill_(dim, index, source);
 }
 
+bool is_radix_sort_accelerated_with_openmp() {
```
Sure, updated the logic to continue using fbgemm::is_radix_sort_accelerated_with_openmp() for fbgemm path.

maajidkhann commented:
> could you also please elaborate on the test plan for this change? do we have tests already exercising this path for AArch64?

**Yes we already have 4 tests covering this patch:**

test/test_scatter_gather_ops.py TestScatterGatherCPU.test_scatter_expanded_index_cpu_float64
test/test_scatter_gather_ops.py TestScatterGatherCPU.test_scatter_expanded_index_cpu_float32
test/test_scatter_gather_ops.py TestScatterGatherCPU.test_scatter_expanded_index_cpu_float16
test/test_scatter_gather_ops.py TestScatterGatherCPU.test_scatter_expanded_index_cpu_bfloat16

maajidkhann commented:
@fadara01 Can you please retrigger the CI pipelines.

maajidkhann committed:
Fix VLA usage in radix_sort_parallel with std::vector

Signed-off-by: maajidkhann <maajidkhan.n@fujitsu.com>

maajidkhann committed:
Fix CI error

mark unused internal functions static to avoid
-Wunneeded-internal-declaration

Signed-off-by: maajidkhann <maajidkhan.n@fujitsu.com>

maajidkhann commented:
The code compiles fine with GCC and even works locally with Clang 16 on Graviton3, but fails only in PyTorch CI, which uses stricter Clang flags. Looks like CI Clang may differ in optimization passes or may be run in C++98/11 compatibility mode where linkage handling is stricter.

Fixed the issue for clang build CI failures.

maajidkhann committed:
Fix static function declarations

- Marked 'combine_prefix_sum' and
'combine_prefix_sum_for_msb' as 'static inline'
to prevent multiple definition errors in header
inclusion.

Signed-off-by: maajidkhann <maajidkhan.n@fujitsu.com>

Skylion007 commented on aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:
```diff
@@ -665,6 +675,208 @@ std::pair<K*, V*> radix_sort_parallel(
+  // work only in debug build.
+  assert(offset == elements_count);
+  // Suppress unused variable warning
+  (void)elements_count;
```
We are on C++17, why not just use [maybe unused] now?

fadara01 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -2008,6 +2008,11 @@ static bool can_use_expanded_index_path(
   if (!fbgemm::is_radix_sort_accelerated_with_openmp()) {
     return false;
   }
+#elif defined(__aarch64__)
```
why is this specific for `__aarch64__`?

maajidkhann replied:
__aarch64__ is used to explicitly gate the optimized path for ARM CPUs. We enable the fast path only when OpenMP is available, as the kernel relies on parallelism. Unlike x86, we don’t have FBGEMM checks here for ARM, so this condition ensures correctness and performance.

So yes, this conditional is both correct and necessary:
1)On ARM architectures (__aarch64__), the optimized radix sort path is enabled only if OpenMP is available, since the kernel relies on OpenMP for parallel execution.

2)For all other architectures (e.g., non-x86, non-ARM), the fast path is conservatively disabled to ensure correctness and portability, as there is no validated fallback or acceleration support in those environments.

maajidkhann committed:
Refactor: use [[maybe_unused]] instead of (void)

Signed-off-by: maajidkhann <maajidkhan.n@fujitsu.com>

maajidkhann commented:
> > done :)
> 
> Thank you. Can we merge this PR now (All the pipelines are passing)

CC @Skylion007 @malfet

maajidkhann commented:
> > > done :)
> > 
> > 
> > Thank you. Can we merge this PR now (All the pipelines are passing)
> 
> CC @Skylion007 @malfet

Hi, can you please review and approve the changes whenever you get time.
Let me know if anything is required from my end. Thanks.

maajidkhann committed:
Addressed review comments

Signed-off-by: maajidkhann <maajidkhan.n@fujitsu.com>

maajidkhann commented:
The one CI failure for CUDA is unrelated to my change. It should pass when we rebase the PR next time.
Please approve the PR for merge!

@fadara01 @Skylion007

fadara01 reviewed (approved):
LGTM, thanks!

maajidkhann commented:
@pytorchbot rebase

maajidkhann commented:
There's just 1 unrelated CUDA pipeline Failure. I see this on other set of PR's as well.
@malfet Can you please review and approve the PR. Thanks.

maajidkhann commented:
@pytorchbot rebase

maajidkhann commented on aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:
```diff
@@ -665,6 +675,204 @@ std::pair<K*, V*> radix_sort_parallel(
+// implementation taken from FBGEMM/src/Utils.cc
+namespace {
+// histogram size per thread
+constexpr int RDX_HIST_SIZE = 256;
```
Done added the explanation @snadampal

maajidkhann commented:
@pytorchbot rebase

aditew01 commented:
@pytorchbot rebase

maajidkhann commented:
@aditew01 All the pipelines now seem to pass.

jgong5 commented on aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:
```diff
@@ -719,13 +929,24 @@ void cpu_scatter_reduce_expanded_index(const Tensor& self, const Tensor& index,
   int64_t* sorted_col_index_keys = nullptr;
   int64_t* sorted_col_index_values = nullptr;
+
+#if defined(USE_FBGEMM)
   std::tie(sorted_col_index_keys, sorted_col_index_values) = fbgemm::radix_sort_parallel(
       keys.get(),
       values.get(),
       keys_tmp.get(),
       values_tmp.get(),
       nnz,
       M);
+#else
+  std::tie(sorted_col_index_keys, sorted_col_index_values) = radix_sort_parallel(
+      keys.get(),
+      values.get(),
+      keys_tmp.get(),
+      values_tmp.get(),
+      nnz,
+      M);
+#endif
```
Do we still need to keep the FBGEMM path if the logic has been copied here?

maajidkhann replied:
@jgong5 Good point — I intentionally kept the USE_FBGEMM branch. Even though the radix-sort logic is implemented in-tree, FBGEMM is still the right place for any x86-specific intrinsic/assembly code in the future and we'd prefer x86 folks to continue landing those optimizations into FBGEMM.

jgong5 replied:
OK, shall we add a comment here to explain this?

maajidkhann replied:
@jgong5 I have added the comment now. Can you please retrigger the CI pipelines.

maajidkhann replied:
Looks like all the checks have passed now.

maajidkhann commented on aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:
```diff
@@ -665,6 +675,206 @@ std::pair<K*, V*> radix_sort_parallel(
+
+  // If negative values are present, we want to perform all passes
+  // up to a sign bit
+  int num_bits = sizeof(K) * 8;
```
@malfet  Thanks for the suggestion! I looked into making these constexpr, but in this function both **num_bits** and **num_passes** depend on the runtime value **max_value**:

Since **countLeadingZeros(max_value**) is evaluated at runtime, **num_bits** cannot be **constexpr** (or const if it needs to be mutated). Likewise, **num_passes** also depends on this runtime-adjusted **num_bits**, so it must remain a runtime constant.

maajidkhann commented on aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:
```diff
@@ -665,6 +675,206 @@ std::pair<K*, V*> radix_sort_parallel(
+    num_bits -= c10::llvm::countLeadingZeros(
+                                 static_cast<typename std::make_unsigned<K>::type>(max_value));
+
+  const unsigned int num_passes = (num_bits + 7) / 8;
```
Same comment as above.

maajidkhann committed:
[aten][cpu] Address reviewer feedback on FBGEMM path

- Added architectural comments clarifying the separation
between FBGEMM (x86) and in-tree (ARM) radix sort paths.

Signed-off-by: maajidkhann <maajidkhan.n@fujitsu.com>

jgong5 commented:
@pytorchbot merge

aditew01 commented:
@malfet @Skylion007 this needs a review, CI LGTM!

maajidkhann commented:
> @malfet @Skylion007 this needs a review, CI LGTM!

Hi @malfet  @Skylion007 - gentle ping on this PR.

This has been in OSS for quite a while now and has already gone through multiple review iterations. CI is green, reviewer feedback has been addressed, and it has an approval from @jgong5 and other reviewers from ARM as well.

It looks like the merge is currently blocked only because a superuser/core maintainer approval is required by the merge bot. Would appreciate it if one of you could take a final look and help get this landed when you have a moment.

Thanks!
